### ISSUE TITLE
disable http2 for certain area or how to use websocket fallback to http1

### ISSUE THREAD

**Issue by ramarro123** [eclipse/jetty.project#2165]
hello,

i would like to know if it's possible to configure jetty to server
https://<>/  with http2 
https://<>/websocket  without http2

basically telling jetty to ignore http2 for "websocket" 

since websocket it's not supported on http2 will be nice if i can have a fallback to http1 in some way, but i don't know how to setup jetty in that way :(

i just embed jetty like that

            org.eclipse.jetty.server.ServerConnector sslConnector = new org.eclipse.jetty.server.ServerConnector(server,
                    new org.eclipse.jetty.server.SslConnectionFactory(sslContextFactory, "http/1.1"),
                    new org.eclipse.jetty.server.SslConnectionFactory(sslContextFactory, "alpn"),
                    new org.eclipse.jetty.alpn.server.ALPNServerConnectionFactory(),
                    new org.eclipse.jetty.http2.server.HTTP2ServerConnectionFactory(sslHttpConfig),
                    new org.eclipse.jetty.server.HttpConnectionFactory(sslHttpConfig));
            sslConnector.setPort(securePort);
            server.addConnector(sslConnector);

i can't figure out if what i want to di it's possible or not :)

**Comment by gregw** [comment]
This is not an issue. Please ask the jetty-user mailing list for such enquiries (or try stack overflow... we do monitor that).

But while I'm here... I'm not exactly sure I understand your question.  I connector configured with http1 and http2 can accept either. If the connection is http1, then it can upgrade to websocket, if it is http2 it cannot.

Note that you will need to configure ALPN, so you accept the right http offered by the client.

**Comment by ramarro123** [comment]
@gregw i found that on firefox the code work, but on chrome no. i'll investigate more and eventually open a new issue with what i found

**Comment by joakime** [comment]
(*total guess*)
Chrome might be attempting the javascript websocket client based on the context/origin of javascript.
The javascript was served in http/2, it could be making an assumption that the attempt to connect back to the same scheme/host/port should use the same HTTP version.

**Comment by ramarro123** [comment]
i just tested it like that:
 i did a page that load in http and open a wss, but i still got the error.
i try to close/restart chrome and still get the invalid http header error, opening wss on the console

so maybe no, it's not that the reason, but i need to do more test.

feel free to give me ideas, i will test them all :D

**Comment by ramarro123** [comment]
Ok, i finish all the tests

i tried on safari, firefox and chrome. 
Only firefox work :(

let me know if i have to open an issue with all the details, but i add them there.

os=osx 10.11
java=jdk1.8.0_66.jdk
jar=alpn-boot-8.1.6.v20151105.jar,jetty-all-9.4.8.v20171121-uber.jar,jetty-alpn-openjdk8-server-9.4.8.v20171121.jar,jetty-alpn-server-9.4.8.v20171121.jar


test code:

    package com.eg.euler.cluster.http.server.jetty;
    
    import org.eclipse.jetty.alpn.server.ALPNServerConnectionFactory;
    import org.eclipse.jetty.http2.server.HTTP2ServerConnectionFactory;
    import org.eclipse.jetty.server.*;
    import org.eclipse.jetty.servlet.DefaultServlet;
    import org.eclipse.jetty.servlet.ServletContextHandler;
    import org.eclipse.jetty.util.ssl.SslContextFactory;
    import org.eclipse.jetty.websocket.api.WebSocketAdapter;
    import org.eclipse.jetty.websocket.servlet.WebSocketServlet;
    import org.eclipse.jetty.websocket.servlet.WebSocketServletFactory;
    
    import java.io.IOException;
    
    public class H2Example {
    
        public static class EchoSocketServlet extends WebSocketServlet {
            @Override
            public void configure(WebSocketServletFactory webSocketServletFactory) {
                webSocketServletFactory.register(EchoSocket.class);
            }
        }
    
        public static class EchoSocket extends WebSocketAdapter {
            @Override
            public void onWebSocketText(String message) {
                try {
                    getRemote().sendString(message);
                } catch (IOException e) {
                    e.printStackTrace(System.err);
                }
            }
    
            @Override
            public void onWebSocketError(Throwable cause) {
                cause.printStackTrace(System.err);
            }
        }
    
        public static void main(String[] args) throws Exception {
            // Setup Jetty Server instance (as jetty.xml)
            Server server = new Server();
    
            int sport = 8443;
            int port = 8080;
    
            server.addBean(new org.eclipse.jetty.util.thread.ScheduledExecutorScheduler());
    
            HttpConfiguration httpConfig = new HttpConfiguration();
            httpConfig.setSecureScheme("https");
            httpConfig.setSecurePort(sport);
    
            org.eclipse.jetty.server.handler.HandlerCollection Handlers = new org.eclipse.jetty.server.handler.HandlerCollection();
            Handlers.setHandlers(new Handler[]{new org.eclipse.jetty.server.handler.ContextHandlerCollection(), new org.eclipse.jetty.server.handler.DefaultHandler()});
            server.setHandler(Handlers);
    
    
            // Setup Jetty Server instance (as jetty-http.xml)
            org.eclipse.jetty.server.ServerConnector httpConnector = new org.eclipse.jetty.server.ServerConnector(server, -1, -1, new HttpConnectionFactory(httpConfig));
            httpConnector.setPort(port);
            server.addConnector(httpConnector);
    
    
            org.eclipse.jetty.server.HttpConfiguration sslHttpConfig = new org.eclipse.jetty.server.HttpConfiguration(httpConfig);
            sslHttpConfig.addCustomizer(new org.eclipse.jetty.server.SecureRequestCustomizer());
    
    
            SslContextFactory sslContextFactory = new SslContextFactory();
            sslContextFactory.setKeyStorePath("rama.jks");
            sslContextFactory.setKeyStorePassword("123456");
    
            ALPNServerConnectionFactory alpn = new ALPNServerConnectionFactory();
            HTTP2ServerConnectionFactory h2 = new HTTP2ServerConnectionFactory(sslHttpConfig);
    
            SslConnectionFactory ssl = new SslConnectionFactory(sslContextFactory, alpn.getProtocol());
    
    
            org.eclipse.jetty.server.ServerConnector sslConnector = new org.eclipse.jetty.server.ServerConnector(server,
                    ssl,
                    alpn,
                    h2,
                    new org.eclipse.jetty.server.HttpConnectionFactory(sslHttpConfig));
            sslConnector.setPort(sport);
    
            server.addConnector(sslConnector);
            sslContextFactory.setUseCipherSuitesOrder(true);
    
    
            ServletContextHandler context = new ServletContextHandler(ServletContextHandler.NO_SECURITY);
            context.setResourceBase("../c3w_test/web/default/test/");
            context.addServlet(DefaultServlet.class, "/").setInitParameter("maxCacheSize", "81920");
    
            server.setHandler(context);
    
            // Add a WebSocket as a Servlet to the context
            context.addServlet(EchoSocketServlet.class, "/echo");
    
    
            server.start();
            server.join();
        }
    }


index page for tests
```
<!DOCTYPE html>
<html>
<head></head>
<body>
<script type="application/javascript">
    var ws = new WebSocket((location.protocol.indexOf('https') >= 0 ? 'wss' : 'ws') + '://192.168.2.123:' + location.port + '/echo');
    ws.onopen = function () {
        for (var i = 0; i < 5; i++)
            ws.send("ping #" + i);
    }
    ws.onclose = function () {
        console.log("aaa");
    }
    ws.onmessage = function (data) {
        console.log(data.data);
    }
    ws.onerror = function () {
        console.log("error");
    }
</script>
test page
</body>
</html>
```


results (on console)

safari+https = WebSocket connection to 'wss://192.168.2.123:8443/echo' failed: Status line contains embedded null
safari+http= ok

chrome+https= VM456:164 WebSocket connection to 'wss://192.168.2.123:8443/echo' failed: Error during WebSocket handshake: net::ERR_INVALID_HTTP_RESPONSE
chrome+http=ok

firefox+https=ok
firefox+http=ok

so basically, it seems to work only on firefox.
client was osx (chrome/firefox/safari) + linux (chrome/firefox)

the versions of browser was the most updated for all platform


hope that that's help, let me know if you prefer a separate issue, or if i can do some more test on my side.

**Comment by joakime** [comment]
Seems you are too bleeding edge for the world of http/2 + websocket.
Perhaps you should put the websocket endpoints on a different port, your javascript can just use that different port without impacting the user any.

**Comment by joakime** [comment]
Pay attention to the draft proposal for WebSocket over HTTP/2 too.
https://datatracker.ietf.org/doc/draft-ietf-httpbis-h2-websockets/

**Comment by joakime** [comment]
Of note: The protocol selection occurs before any knowledge of what is being requested.

In other words, the choice of HTTP/1.1 or HTTP/2 is made before the details of the request arrive.
The server has already selected the protocol way before it even knows that this request is even an Upgrade to WebSocket.
